### PR TITLE
Avoid unnecessarily creating numpy array of Var objects

### PR DESCRIPTION
**Pull request by simonbowly**
The array is unused in the code path for any currently supported version of Gurobi (>= v9.5) so can be removed to avoid unnecessary overhead. Additionally, in Python 3.12, creating this array triggers a [memory leak due to a bug in gurobipy](https://support.gurobi.com/hc/en-us/articles/27635574446353-Memory-leak-in-gurobipy-with-Python-3-12), so it would be good to remove it from cvxpy to reduce user impact.

Figured I'd start with a minimal change, but I'm happy to do a larger rewrite to remove the old workarounds in `gurobi_qpif.py`.

## Description
Please include a short summary of the change.
Issue link (if applicable):

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [x] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [x] Run the unittests and check that they’re passing (`pytest cvxpy/tests/test_qp_solvers.py` passes with gurobipy v11.0.3)
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

**Review by Transurgeon — COMMENTED** [review]
LGTM. also gonna add a backport tag to this PR.

File: cvxpy/reductions/solvers/qp_solvers/gurobi_qpif.py [hunk 1]
```diff
@@ -181,7 +181,12 @@ def solve_via_data(self, data, warm_start: bool, verbose: bool, solver_opts, sol
                 x_grb[idx].start = data['init_value'][idx]
         model.update()
 
-        x = np.asarray(model.getVars())
+        if hasattr(model, 'addMConstr') and hasattr(model, 'setMObjective'):
```
Comment (PTNobel) [inline cvxpy/reductions/solvers/qp_solvers/gurobi_qpif.py:184]: Could we just check the version number?

**Review by PTNobel — COMMENTED** [review]
If it's not needed for anything we support, any reason to not just delete the code?

File: cvxpy/reductions/solvers/qp_solvers/gurobi_qpif.py (hunk 1 above)
Comment (simonbowly) [reply cvxpy/reductions/solvers/qp_solvers/gurobi_qpif.py:184]: Sure, but I would go one way or the other for all the following branches.

**Comment by simonbowly** [comment]
> If it's not needed for anything we support, any reason to not just delete the code?

I should clarify: by "supported" I mean versions that Gurobi still officially supports. As of now that's gurobi 9.5 and later. I'm not sure what cvxpy still intends to support in terms of Gurobi versions.

**Comment by PTNobel** [comment]
> I'm not sure what cvxpy still intends to support in terms of Gurobi versions.

We only support what Gurobi supports is our policy as of today.

**Comment by Transurgeon** [comment]
@simonbowly To move forward with this PR, I propose that we delete the old code since cvxpy only supports the same versions as Gurobi. What do you think?

**Comment by simonbowly** [comment]
> To move forward with this PR, I propose that we delete the old code since cvxpy only supports the same versions as Gurobi. What do you think?

Works for me! This is now done with deafe23. The tests pass against gurobipy 9.5.2, 10.0.3, and 11.0.3.

Note that I also removed the update() calls. They're no longer needed in between adding variables and constraints in recent versions of Gurobi.

**Review by Transurgeon — APPROVED** [review]
LGTM.

**Review by PTNobel — APPROVED** [review]

**Comment by PTNobel** [comment]
Thank you! I set the CI to run, should be good to merge once it passes!

**Pull request merged**